pom.xml

```diff
@@ -1830,7 +1851,7 @@
                 <plugin>
                     <groupId>org.apache.maven.plugins</groupId>
                     <artifactId>maven-dependency-plugin</artifactId>
-                    <version>3.1.1</version>
+                    <version>3.1.2</version>
```

Comment:
I tried exploring different versions for this.

If we don't explicitly specify a version, it uses `3.2.0` from the parent, which runs into the following issue:
```
<pmd xmlns="http://pmd.sourceforge.net/report/2.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://pmd.sourceforge.net/report/2.0.0 http://pmd.sourceforge.net/report_2_0_0.xsd" version="6.42.0" timestamp="2024-11-18T17:56:13.310">
<error filename="/Users/akjn/code/druid/processing/src/main/java/org/apache/druid/query/aggregation/JavaScriptAggregatorFactory.java" msg="PMDException: Error while parsing /Users/akjn/code/druid/processing/src/main/java/org/apache/druid/query/aggregation/JavaScriptAggregatorFactory.java">
<![CDATA[ net.sourceforge.pmd.PMDException: Error while parsing /Users/akjn/code/druid/processing/src/main/java/org/apache/druid/query/aggregation/JavaScriptAggregatorFactory.java at net.sourceforge.pmd.SourceCodeProcessor.processSourceCodeWithoutCache(SourceCodeProcessor.java:124) at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:100) at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:62) at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:85) at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:29) at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) at java.base/java.lang.Thread.run(Thread.java:840) Caused by: net.sourceforge.pmd.lang.java.ast.ParseException: Encountered " "." ". "" at line 71, column 40. Was expecting one of: <IDENTIFIER> ... "@" ... "@" ... "@" ... "@" ... <IDENTIFIER> ... <IDENTIFIER> ... <IDENTIFIER> ... <IDENTIFIER> ... <IDENTIFIER> ... <IDENTIFIER> ... <IDENTIFIER> ... <IDENTIFIER> ... <IDENTIFIER> ... <IDENTIFIER> ... <IDENTIFIER> ... <IDENTIFIER> ... <IDENTIFIER> ... <IDENTIFIER> ... <IDENTIFIER> ... at net.sourceforge.pmd.lang.java.ast.JavaParser.generateParseException(JavaParser.java:13532) at net.sourceforge.pmd.lang.java.ast.JavaParser.jj_consume_token(JavaParser.java:13416) at net.sourceforge.pmd.lang.java.ast.JavaParser.MethodDeclarator(JavaParser.java:2169) at net.sourceforge.pmd.lang.java.ast.JavaParser.MethodDeclaration(JavaParser.java:2118) at net.sourceforge.pmd.lang.java.ast.JavaParser.ClassOrInterfaceBodyDeclaration(JavaParser.java:1811) at net.sourceforge.pmd.lang.java.ast.JavaParser.ClassOrInterfaceBody(JavaParser.java:1764) at net.sourceforge.pmd.lang.java.ast.JavaParser.ClassOrInterfaceDeclaration(JavaParser.java:905) at net.sourceforge.pmd.lang.java.ast.JavaParser.TypeDeclaration(JavaParser.java:807) at net.sourceforge.pmd.lang.java.ast.JavaParser.CompilationUnit(JavaParser.java:527) at net.sourceforge.pmd.lang.java.AbstractJavaParser.parse(AbstractJavaParser.java:62) at net.sourceforge.pmd.lang.AbstractParser.doParse(AbstractParser.java:44) at net.sourceforge.pmd.SourceCodeProcessor.parse(SourceCodeProcessor.java:136) at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:200) at net.sourceforge.pmd.SourceCodeProcessor.processSourceCodeWithoutCache(SourceCodeProcessor.java:118) ... 10 more ]]>
</error>
</pmd>
```

The above can be simulated using:
```
mvn dependency:analyze -DoutputXML=true -DignoreNonCompile=true -DfailOnWarning=true -X 
```

The above is happening for the following line: https://github.com/apache/druid/blob/24a1fafaa745fb93bf1522490c095cd689fd9996/processing/src/main/java/org/apache/druid/query/aggregation/JavaScriptAggregatorFactory.java#L71

I tried using `@SuppressWarnings()` for the above line, but that also doesn't help since it's a parsing error for PMD.

Thoughts?

Reply:
I've removed the version setting for the pmd plugin and it started working; it found a few new warnings - but we could possibly disable the checks for those for now?

Reply:
I explored further on this, have the following possible options:

### Option 1

- Remove version from maven-dependency-plugin, hence it defaults to 3.2.0.
- Set version of maven-pmd-plugin as 3.26.0 (this is the latest version).
- Remove `<rule ref="category/java/codestyle.xml/UnnecessaryFullyQualifiedName"/>` from [pmd-ruleset.xml](https://github.com/apache/druid/blob/master/codestyle/pmd-ruleset.xml) (because there were hundreds of errors)

- With the above changes, we get a LOT of warnings (hence the check fails) like:
```
[WARNING] Non-test scoped test only dependencies found:
[WARNING]    org.apache.zookeeper:zookeeper-jute:jar:3.8.4:compile
[WARNING]    commons-codec:commons-codec:jar:1.16.1:compile
[WARNING]    joda-time:joda-time:jar:2.12.5:compile
[WARNING]    org.apache.curator:curator-client:jar:5.5.0:compile
```

To address the above, I added `<scope>provided</scope>` to the dependencies mentioned in the above warning message. I did that in 5 modules, but since this would have to be done manually for (most likely) almost all 80 modules, I thought we should discuss first if we want to go with this approach.

### Option 2

- Set maven-dependency-plugin version to 3.3.0.
  - This allows us to add a configuration to pom.xml to ignore these type of errors, so add the following:
```
                    <configuration>
                        <ignoredNonTestScopedDependencies>
                            <ignoredNonTestScopedDependency>*</ignoredNonTestScopedDependency>
                        </ignoredNonTestScopedDependencies>
                    </configuration>
```
- Set version of maven-pmd-plugin as 3.26.0 (this is the latest version).
- Remove `<rule ref="category/java/codestyle.xml/UnnecessaryFullyQualifiedName"/>` from [pmd-ruleset.xml](https://github.com/apache/druid/blob/master/codestyle/pmd-ruleset.xml) (because there were hundreds of errors)

### Option 3

- Keep the existing PR as it is. This means:
  - maven-dependency-plugin == 3.1.2
  - maven-pmd-plugin == 3.16.0
- Once this PR is merged, have a follow-up work item, where we proceed with either option 1 or option 2. In that follow-up PR, we can also work on the `UnnecessaryFullyQualifiedName` violations, instead of removing this from `pmd-ruleset.xml`

<hr>

My thoughts: Option 1 would affect a lot of files, and make it difficult to review the current PR. So I feel that we should either go with Option 2 in this PR, or go with Option 3.

What do you think? Would appreciate your inputs on the above, thanks!

Reply:
> note: I'm not sure why - but somehow I don't see Non-test scoped test only dependencies found issue right now locally

@kgyrtkirk It's because we get those issues on updating `maven-pmd-plugin` to the latest version `3.26.0`

Reply:
> anyway: Option 2 sounds like a nice way of mitigating that...I always feeled that its pointless to declare these for tests as well...

Have pushed the change to the PR for option 2.